onnxruntime/core/providers/cuda/tunable/math/gemm.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/tunable/math/gemm.h"
+
+#include "core/common/inlined_containers.h"
+
+namespace onnxruntime {
+namespace cuda {
+namespace tunable {
+
+template <typename T>
+GemmParams<T>::GemmParams(int m, int n, int k, bool trans_a, bool trans_b, float alpha, float beta,
+                          const Gemm<T>* gemm_kernel, OpKernelContext* ctx)
+    : OpParams(gemm_kernel->GetTuningContext(), gemm_kernel->Stream(ctx)),
+      trans_a_(trans_a),
+      trans_b_(trans_b),
+      alpha_(alpha),
+      beta_(beta),
+      m_(m),
+      n_(n),
+      k_(k),
+      gemm_kernel_(gemm_kernel),
+      ctx_(ctx) {
+  const auto* b = ctx->Input<Tensor>(2);
+  bm_ = gsl::narrow_cast<int>(beta_ == 0.0f ? 0 : (b->Shape().NumDimensions() > 1 ? b->Shape()[0] : 1));
+  bn_ = gsl::narrow_cast<int>(
+      beta_ == 0.0f
+          ? 0
+          : (b->Shape().NumDimensions() > 1 ? b->Shape()[1] : (b->Shape().NumDimensions() > 0 ? b->Shape()[0] : 1)));
+
+#ifdef ENABLE_TRITON
+  const auto* x = ctx->Input<Tensor>(0);
+  has_triton_support_ = contrib::IsTritonOpExecutorInitialized() &&
+                        (std::is_same<T, MLFloat16>::value || std::is_same<T, float>::value) &&
+                        x->Shape().NumDimensions() > 1;
+#endif
+}
+
+namespace {
+
+template <typename T>
+common::Status DefaultGemmOp(const GemmParams<T>* params) {
+  return params->gemm_kernel_->ComputeDefault(params->ctx_, params->m_, params->n_, params->k_);
+}
+
+#ifdef ENABLE_TRITON
+template <typename T>
+common::Status TritonGemmOp(const GemmParams<T>* params) {
+  TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(!params->has_triton_support_);
```

Comment:
I am curious, in what case triton is enabled but there is no triton support?

Reply:
The original Gemm Op can support x as vector, but I don't handle such case in triton code. And currently triton code supports float and bf16 type only.